### PR TITLE
attempt the pinning workflow

### PR DESCRIPTION
**Pull request by stan-dot**

### CODE REVIEW

**Comment by stan-dot** [comment]
note: the change will not be checked on this branch, as release only runs on main

**Comment by coretl** [comment]
The suggested workflow in https://diamondlightsource.github.io/python-copier-template/main/how-to/lock-requirements.html is this:
- do the `pip freeze` locally to make `dev-requirements.txt`
- check it in
- the container build uses this as a constraints file

Will this cover what you're trying to do?

**Comment by stan-dot** [comment]
@coretl you mean this `_container`? file? `      - name: Set up Docker Buildx
        id: buildx
        uses: docker/setup-buildx-action@v3

      - name: Log in to GitHub Docker Registry
        if: github.event_name != 'pull_request'
        uses: docker/login-action@v3
        with:
          registry: ghcr.io
          username: ${{ github.actor }}
          password: ${{ secrets.GITHUB_TOKEN }}

      - name: Build and export to Docker local cache
        uses: docker/build-push-action@v5
        with:
          context: .
          # Need load and tags so we can test it below
          load: true
          tags: tag_for_testing`
          
         I'm not sure where this file would go?

**Comment by coretl** [comment]
I'm now confused, I thought we changed python-copier-template so this line:
https://github.com/DiamondLightSource/blueapi/blob/54b04966ef3ee48f6774122954f1c01b08ff7200/Dockerfile#L19
would be:
```dockerfile
RUN touch dev-requirements.txt && pip install -c dev-requirements.txt .
```
That would use the dev-requirements.txt (if checked into the repo) as a set of constraints, effectively pinning the container dependencies in:
https://github.com/DiamondLightSource/blueapi/blob/54b04966ef3ee48f6774122954f1c01b08ff7200/.github/workflows/_container.yml#L27-L33

Please can you make that change locally to blueapi and see if your pinned dev-dependencies end up in the built container?

**Comment by stan-dot** [comment]
neither my branch nor main have that line in the Dockerfile. 

building now

**Comment by DiamondJoseph** [comment]
https://github.com/DiamondLightSource/python-copier-template/blob/8c0033b9bb3dff48f955fe2bae7045a5a4e10736/template/Dockerfile.jinja#L19

Doesn't appear to be in the copier template

**Comment by stan-dot** [comment]
> 
Run pip install $([ -f dev-requirements.txt ] && echo '-c dev-requirements.txt') -e .[dev]
DEPRECATION: Constraints are only allowed to take the form of a package name and a version specifier. Other forms were originally permitted as an accident of the implementation, but were undocumented. The new implementation of the resolver no longer supports these forms. A possible replacement is replacing the constraint with a requirement. Discussion can be found at https://github.com/pypa/pip/issues/8210

the linter does not like this way of accomplishing the goal @coretl

**Comment by stan-dot** [comment]
> Run docker run --rm tag_for_testing --version
> .blueapi scratch area not present
> /container-startup.sh: line 19: blueapi: command not found
> Error: Process completed with exit code 127.

install without the editable causes an error -

File: Dockerfile [hunk 1]
```diff
@@ -16,7 +16,7 @@ ENV PATH=/venv/bin:$PATH
 FROM developer as build
 COPY . /context
 WORKDIR /context
-RUN pip install .
+RUN touch dev-requirements.txt && pip install --upgrade pip && pip install -r dev-requirements.txt
```
Comment (coretl) [inline Dockerfile:19]: As I said in https://github.com/DiamondLightSource/blueapi/pull/499#issuecomment-2158961600

please make this:
```suggestion
RUN touch dev-requirements.txt && pip install --upgrade pip && pip install -c dev-requirements.txt .
```
```

**Comment by stan-dot** [comment]
> 
=========================== short test summary info ============================
FAILED tests/core/test_context.py::test_add_devices_from_module - ValueError: too many values to unpack (expected 2)
FAILED tests/core/test_context.py::test_add_failing_deivces_from_module - TimeoutError: FooBar
FAILED tests/core/test_context.py::test_add_devices_and_plans_from_modules_with_config - ValueError: too many values to unpack (expected 2)
FAILED tests/messaging/test_stomptemplate.py::test_messaging_template_can_be_set_with_none - TimeoutError: This device is supposed to fail, blueapi will mark it as not present and start up regardless
FAILED tests/service/test_rest_api.py::test_teardown_handler - TimeoutError: This device is supposed to fail, blueapi will mark it as not present and start up regardless
FAILED tests/service/test_subprocess_handler.py::test_initialize - TimeoutError: This device is supposed to fail, blueapi will mark it as not present and start up regardless
FAILED tests/service/test_subprocess_handler.py::test_reload - TimeoutError: This device is supposed to fail, blueapi will mark it as not present and start up regardless
FAILED tests/service/test_subprocess_handler.py::test_method_routing - pytest.PytestUnraisableExceptionWarning: Exception ignored in: <function Pool.__del__ at 0x7f9dc3439620>

hmm

File: .github/workflows/_release.yml
Comment (coretl) [inline .github/workflows/_release.yml:1]: Please revert this entire file

**Review by coretl — CHANGES_REQUESTED** [review]

**Comment by coretl** [comment]
So these tests are now doing what is expected:
- `CI / test (ubuntu-latest, 3.11)` passes, as it uses the latest version of `ophyd-async`
- `CI / test (ubuntu-latest, dev)` fails, as it uses the version you have specified in your lockfile, which is out of date

Please delete `dev-requirements.txt`, restart your dev container, check the tests run, then run `pip freeze --exclude-editable > dev-requirements.txt` and check it in.

If this works, please can you also add these instructions to a new "Updating the lockfile" section in https://diamondlightsource.github.io/python-copier-template/main/how-to/lock-requirements.html

**Comment by stan-dot** [comment]
> Please delete dev-requirements.txt, restart your dev container, check the tests run, then run pip freeze --exclude-editable > dev-requirements.txt and check it in.

for me that's a great use case for a makefile to avoid all that typing, and typos.

**Review by coretl — APPROVED** [review]

**Comment by coretl** [comment]
@DiamondJoseph feel free to merge if you're happy with this

**Review by DiamondJoseph — CHANGES_REQUESTED** [review]
Blocking submit while I have a look

**Comment by stan-dot** [comment]
@DiamondJoseph what are your thoughts?

**Comment by joeshannon** [comment]
Neither the commit message or the PR description provide a motivation for this change - would you mind adding it for future reference?

Does this change the procedure for making a dev environment?

I've just tried with `pip install -e .[dev]` and have ended up with newer versions of dependencies than those referenced in `dev-requirements.txt`.

**Comment by stan-dot** [comment]
looks like you'd need to then run `pip install -r dev-requirements.txt` for it to work. I need to think about adding this to the docs

**Pull request merged**